File: bindings/js-browser/test/chat.test.js
/**
 * chat.test.js – unit tests for LlamaChat (browser binding).
 *
 * Tests run against the stub WASM module so no model file is required.
 */

import { LlamaChat } from '../src/chat.js';
import { LlamaError } from '../src/errors.js';

describe('LlamaChat', () => {

  test('load returns a LlamaChat instance', async () => {
    const chat = await LlamaChat.load('dummy.gguf');
    expect(chat).toBeInstanceOf(LlamaChat);
    chat.close();
  });

  test('load with empty path throws MODEL_LOAD_FAILED', async () => {
    await expect(LlamaChat.load('')).rejects.toThrow(LlamaError);
  });

  test('chat text mode returns assistant_text', async () => {
    const chat = await LlamaChat.load('dummy.gguf');
    try {
      const resp = await chat.chat({
        messages: [{ role: 'user', content: 'Say hello.' }],
      });
      expect(resp.type).toBe('assistant_text');
      expect(typeof resp.text).toBe('string');
    } finally {
      chat.close();
    }
  });

  test('chat json_schema mode returns structured_json', async () => {
    const chat = await LlamaChat.load('dummy.gguf');
    try {
      const resp = await chat.chat({
        messages:     [{ role: 'user', content: 'Extract data.' }],
        responseMode: 'json_schema',
        schema:       { name: 'person', schema: { type: 'object' } },
      });
      expect(resp.type).toBe('structured_json');
    } finally {
      chat.close();
    }
  });

  test('chat tool_call mode returns tool_call', async () => {
    const chat = await LlamaChat.load('dummy.gguf');
    try {
      const resp = await chat.chat({
        messages:     [{ role: 'user', content: 'Weather in Chennai?' }],
        responseMode: 'tool_call',
        tools: [{ name: 'lookup_weather', description: 'Get weather', parameters: {} }],
        toolChoice: 'auto',
      });
      expect(resp.type).toBe('tool_call');
      expect(Array.isArray(resp.tool_calls)).toBe(true);
    } finally {
      chat.close();
    }
  });

  test('chat after close throws ENGINE_CLOSED', async () => {
    const chat = await LlamaChat.load('dummy.gguf');
    chat.close();
    await expect(
      chat.chat({ messages: [{ role: 'user', content: 'hi' }] })
    ).rejects.toMatchObject({ code: 'ENGINE_CLOSED' });
  });

  test('close is idempotent', async () => {
    const chat = await LlamaChat.load('dummy.gguf');
    chat.close();
    expect(() => chat.close()).not.toThrow();
  });

  test('onEvent callback receives events', async () => {
    const events = [];
    const chat = await LlamaChat.load('dummy.gguf', {
      onEvent: (e) => events.push(e),
    });
    chat.close();
    // Load events are emitted by the bridge stub; at minimum we verify
    // that the callback wiring doesn't throw.
    expect(Array.isArray(events)).toBe(true);
  });

  test('missing messages returns error', async () => {
    const chat = await LlamaChat.load('dummy.gguf');
    try {
      await expect(
        chat.chat({ messages: [] })
      ).rejects.toThrow(LlamaError);
    } finally {
      chat.close();
    }
  });
});

File: bindings/go/llama/chat_test.go
package llama_test

import (
	"testing"

	"github.com/muthuishere/llama-bindings/go/llama"
)

// TestChatEngineCloseIdempotent verifies that calling Close twice does not panic.
func TestChatEngineCloseIdempotent(t *testing.T) {
	chat, err := llama.NewChat("testdata/dummy.gguf", llama.LoadOptions{})
	if err != nil {
		t.Skipf("model not available: %v", err)
	}
	chat.Close()
	chat.Close() // must not panic
}

// TestChatAfterClose returns ENGINE_CLOSED error.
func TestChatAfterClose(t *testing.T) {
	chat, err := llama.NewChat("testdata/dummy.gguf", llama.LoadOptions{})
	if err != nil {
		t.Skipf("model not available: %v", err)
	}
	chat.Close()

	_, err = chat.Chat(llama.ChatRequest{
		Messages: []llama.Message{{Role: "user", Content: "hello"}},
	}, llama.ChatOptions{})

	if err == nil {
		t.Fatal("expected error after close, got nil")
	}
	ll, ok := err.(*llama.LlamaError)
	if !ok {
		t.Fatalf("expected *llama.LlamaError, got %T", err)
	}
	if ll.Code != llama.ErrCodeEngineClosed {
		t.Fatalf("expected ENGINE_CLOSED, got %s", ll.Code)
	}
}

// TestChatInvalidModelPath verifies NewChat returns an error for a bad path.
func TestChatInvalidModelPath(t *testing.T) {
	_, err := llama.NewChat("/nonexistent/model.gguf", llama.LoadOptions{})
	if err == nil {
		t.Fatal("expected error for invalid model path")
	}
}

// TestChatMissingMessages verifies that an empty Messages slice is rejected.
func TestChatMissingMessages(t *testing.T) {
	chat, err := llama.NewChat("testdata/dummy.gguf", llama.LoadOptions{})
	if err != nil {
		t.Skipf("model not available: %v", err)
	}
	defer chat.Close()

	_, err = chat.Chat(llama.ChatRequest{}, llama.ChatOptions{})
	if err == nil {
		t.Fatal("expected error for missing messages")
	}
}

// TestChatTextMode exercises the text response mode.
func TestChatTextMode(t *testing.T) {
	chat, err := llama.NewChat("testdata/dummy.gguf", llama.LoadOptions{})
	if err != nil {
		t.Skipf("model not available: %v", err)
	}
	defer chat.Close()

	resp, err := chat.Chat(llama.ChatRequest{
		Messages: []llama.Message{
			{Role: "user", Content: "Say hello in one short sentence."},
		},
		ResponseMode: llama.ResponseModeText,
		Generation: llama.GenerationOptions{
			Temperature:     0.2,
			MaxOutputTokens: 64,
		},
	}, llama.ChatOptions{})
	if err != nil {
		t.Fatalf("chat error: %v", err)
	}
	if resp.Type != "assistant_text" {
		t.Fatalf("expected assistant_text, got %s", resp.Type)
	}
	if resp.Text == "" {
		t.Fatal("expected non-empty text response")
	}
}

// TestChatSchemaMode exercises the json_schema response mode.
func TestChatSchemaMode(t *testing.T) {
	chat, err := llama.NewChat("testdata/dummy.gguf", llama.LoadOptions{})
	if err != nil {
		t.Skipf("model not available: %v", err)
	}
	defer chat.Close()

	resp, err := chat.Chat(llama.ChatRequest{
		Messages: []llama.Message{
			{Role: "user", Content: "John is 32 years old. Extract structured data."},
		},
		ResponseMode: llama.ResponseModeJSONSchema,
		Schema: &llama.Schema{
			Name: "person_extract",
			Schema: map[string]interface{}{
				"type": "object",
				"properties": map[string]interface{}{
					"name": map[string]interface{}{"type": "string"},
					"age":  map[string]interface{}{"type": "integer"},
				},
				"required":             []string{"name", "age"},
				"additionalProperties": false,
			},
		},
		Generation: llama.GenerationOptions{Temperature: 0.0, MaxOutputTokens: 64},
	}, llama.ChatOptions{})
	if err != nil {
		t.Fatalf("chat error: %v", err)
	}
	if resp.Type != "structured_json" {
		t.Fatalf("expected structured_json, got %s", resp.Type)
	}
}

// TestChatToolCallMode exercises the tool_call response mode.
func TestChatToolCallMode(t *testing.T) {
	chat, err := llama.NewChat("testdata/dummy.gguf", llama.LoadOptions{})
	if err != nil {
		t.Skipf("model not available: %v", err)
	}
	defer chat.Close()

	resp, err := chat.Chat(llama.ChatRequest{
		Messages: []llama.Message{
			{Role: "user", Content: "What is the weather in Chennai?"},
		},
		ResponseMode: llama.ResponseModeToolCall,
		Tools: []llama.ToolDefinition{
			{
				Name:        "lookup_weather",
				Description: "Get weather by city",
				Parameters: map[string]interface{}{
					"type": "object",
					"properties": map[string]interface{}{
						"city": map[string]interface{}{"type": "string"},
					},
					"required":             []string{"city"},
					"additionalProperties": false,
				},
			},
		},
		ToolChoice: "auto",
		Generation: llama.GenerationOptions{Temperature: 0.2, MaxOutputTokens: 64},
	}, llama.ChatOptions{})
	if err != nil {
		t.Fatalf("chat error: %v", err)
	}
	if resp.Type != "tool_call" {
		t.Fatalf("expected tool_call, got %s", resp.Type)
	}
}

// TestChatEventCallback verifies that load and chat events are emitted.
func TestChatEventCallback(t *testing.T) {
	var events []string
	onEvent := func(e llama.Event) {
		events = append(events, e.Event)
	}

	chat, err := llama.NewChat("testdata/dummy.gguf", llama.LoadOptions{OnEvent: onEvent})
	if err != nil {
		t.Skipf("model not available: %v", err)
	}
	defer chat.Close()

	_, _ = chat.Chat(llama.ChatRequest{
		Messages: []llama.Message{{Role: "user", Content: "hi"}},
	}, llama.ChatOptions{})

	if len(events) == 0 {
		t.Fatal("expected at least one event, got none")
	}
}
